### PR TITLE
Add DiagnosticSuppressor to suppress the IDE0002 Style Hint in the IDE

### PR DESCRIPTION
**Pull request by dellis1972**
Context #8381

Add a `DiagnosticSuppressor` to "turn off" the `IDE0002` style diagnostic message which incorrectly tells users to use the `_Microsoft.Android.Resource.Designer.ResourceConstant` type directly. This can cause allot of annoyance with our users because it appears on EVERY single usage of `Resource.*`. So you end up with what looks like code spagetti. 

So we need to start shipping an `Analyzer` assembly along with the `Ref` framework pack. This is the place these things need to go. Unfortunately it means that the older frameworks will not get this analyzer. Only the current one. 

On the packaging side, the Analyzer assembly has to go in a `analyzers/dotnet/<language>` folder in the .Ref Nuget Package. There also needs to be an entry in the `FrameworkList.xml` file which has a `Type="Analyzer" ` and a `Language="cs"`. This allows the IDE's to pickup the code. We can ship both regular Analyzers and the DiagnosticSuppressors in the same assembly. So we can extend this with more if needed.

How this works is we use Rosyln to look for the IDE0002 diagnsotic message, we then check the code and see if it is refering to a `_Microsoft.Android.Resource.Designer.*` derived class. If it is , we add a suppression. This will stop the hint appearing in the IDE. I have tested this on VS in devbox and it appears to work . 

Also we generate a Resource Designer assembly and an Intermediate source file at build time. Both of these contain classes which should have the `GeneratedCode` Attribute. So lets add it. The version will be the same as the build assembly used to generate it.

### CODE REVIEW

File: src/Xamarin.Android.Build.Tasks/Tasks/GenerateResourceDesignerIntermediateClass.cs [hunk 1]
```diff
@@ -21,13 +21,15 @@ public class GenerateResourceDesignerIntermediateClass : AndroidTask
...
 using System;
+using System.CodeDom.Compiler;
 
 namespace %NAMESPACE% {
 	#pragma warning disable IDE0002
 	/// <summary>
 	/// Android Resource Designer class.
 	/// Exposes the Android Resource designer assembly into the project Namespace.
 	/// </summary>
+	[GeneratedCode(""%TOOL%"", ""%VERSION%"")]
```
Comment (jonathanpeppers) [inline src/Xamarin.Android.Build.Tasks/Tasks/GenerateResourceDesignerIntermediateClass.cs:32]: Does this need to do:
```suggestion
	[global::System.CodeDom.Compiler.GeneratedCode(""%TOOL%"", ""%VERSION%"")]
```
I'm just thinking if someone makes their own attribute named `GeneratedCode`.

**Review by jonathanpeppers — COMMENTED** [review]
I guess we need to update the `.apkdesc` files:
```
File 'lib/arm64-v8a/lib__Microsoft.Android.Resource.Designer.dll.so' has changed by 87 bytes (7.24 %).
```
I think the threshold is like 5% and this is a really small file.

File: src/Microsoft.Android.Sdk.Analysis/ResourceDesignerDiagnosticSuppressor.cs [hunk 2]
```diff
@@ -0,0 +1,71 @@
+using System.Linq;
+using System.Collections.Immutable;
+using System.Threading;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Text;
+
+[DiagnosticAnalyzer(LanguageNames.CSharp)]
+public class ResourceDesignerDiagnosticSuppressor : DiagnosticSuppressor
```
Comment (jonathanpeppers) [inline src/Microsoft.Android.Sdk.Analysis/ResourceDesignerDiagnosticSuppressor.cs:10]: When I was writing this:
* https://github.com/jonathanpeppers/memory-analyzers

The project template for a Roslyn analyzer created a unit test project:
* https://github.com/jonathanpeppers/memory-analyzers/tree/main/MemoryAnalyzers/MemoryAnalyzers.Test

Would the same test project work here? We could do it in a future PR.

I'm thinking there are other things we could add to this in the future like:
* https://github.com/dotnet/android/wiki/Potential-Roslyn-Analyzers

**Review by jonathanpeppers — APPROVED** [review]

File: src/Microsoft.Android.Sdk.Analysis/ResourceDesignerDiagnosticSuppressor.cs (hunk 2 above)
Comment (dellis1972) [reply src/Microsoft.Android.Sdk.Analysis/ResourceDesignerDiagnosticSuppressor.cs:10]: I think something like that for a future PR would be good. And this new project is a good place for analyzers as well as suppressors.

**Pull request merged**